packages/block-library/src/gallery/edit.js

```diff
@@ -125,11 +125,42 @@ class GalleryEdit extends Component {
 		};
 	}
 
-	onSelectImages( images ) {
-		const { columns } = this.props.attributes;
+	selectCaption( newImage, images, attachmentCaptions ) {
+		const currentImage = find(
+			images, { id: newImage.id }
+		);
+
+		const currentImageCaption = currentImage ? currentImage.caption : newImage.caption;
+
+		if ( ! attachmentCaptions ) {
+			return currentImageCaption;
+		}
+
+		const attachment = find(
+			attachmentCaptions, { id: newImage.id }
+		);
+
+		// if the attachment caption is updated
+		if ( attachment && ( attachment.caption !== newImage.caption ) ) {
+			return newImage.caption;
+		}
+
+		return currentImageCaption;
+	}
+
+	onSelectImages( newImages ) {
+		const { columns, images, attachmentCaptions } = this.props.attributes;
+
 		this.setAttributes( {
-			images: images.map( ( image ) => pickRelevantMediaFiles( image ) ),
-			columns: columns ? Math.min( images.length, columns ) : columns,
+			images: newImages.map( ( newImage ) => ( {
+				...pickRelevantMediaFiles( newImage ),
+				caption: this.selectCaption( newImage, images, attachmentCaptions ),
+			} ) ),
+			attachmentCaptions: newImages.map( ( newImage ) => ( {
```

Comment:
We discussed the current solution during the pairing session earlier today. Randomly failing tests on Travis and in local env are a bit puzzling. There is this question whether those captions could be stored in component's state instead?

Reply:
Moved it to state instead of attributes and so the need to edit block.json is gone. Let's see how this tests @gziolo